### PR TITLE
[Prim] Implement group_norm_backward

### PR DESCRIPTION
**Pull request by malfet**
Test plan: CI, i.e. `python3 test_decomp.py -v -k test_comprehensive_nn_functional_group_norm` plus:
```
#!/usr/bin/env python3.8
import torch

func = torch.ops.aten.native_group_norm_backward.default
decomp =  torch._decomp.decomposition_table[func]
for args in (
        (torch.rand(1, 6, 3), torch.rand(1, 6, 3), torch.rand(1, 2), torch.rand(1, 2), torch.rand(6), 1, 6, 3, 2, [True, True, True]),
        (torch.rand(64, 768, 7, 7), torch.rand(64, 768, 7, 7), torch.rand(64, 1), torch.rand(64, 1), torch.rand(768), 64, 768, 49, 1, [True, True, True])):
    nrc=func(*args)
    drc=decomp(*args)
    for i in range(len(nrc)):
       print(i, torch.max(nrc[i]-drc[i]))
    print(all(torch.allclose(x, y) for (x, y) in zip(nrc, drc)))
```

@diff-train-skip-merge

### CODE REVIEW

**Comment by Chillee** [comment]
I think you actually do fail CUDA tests for float16/bfloat16 accuracy :) The lack of tests for masks other than `[True, False, False]` is ... unfortunate though.

**Comment by malfet** [comment]
> I think you actually do fail CUDA tests for float16/bfloat16 accuracy :) The lack of tests for masks other than `[True, False, False]` is ... unfortunate though.

Good point, haven't realized that float16 tests on CPU are no-ops (say should have been reported as skipped then, or may be not even generated)

Current iteration should be clean on CUDA, will  extend the test to cover other masks

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+    output_mask: List[bool],
+) -> Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor]]:
+    assert grad_output.shape == input.shape
+    assert input.numel() == N * C * HxW
+    assert mean.shape == (N, group)
+    assert rstd.shape == (N, group)
+    assert gamma is None or gamma.numel() == C
+
+    # Compute Internal gradients
+    acc_dtype = utils.get_computation_dtype(input.dtype)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:870]: probably do @pw_cast_for_opmath instead of this, to make sure all inputs are converted

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+    ds = torch.mul(grad_output, input.to(dtype=acc_dtype)).view(N, C, HxW).sum(dim=[2])
+    db = grad_output.view(N, C, HxW).sum(dim=[2], dtype=acc_dtype)
+
+    d_input: Optional[Tensor] = None
+    d_gamma: Optional[Tensor] = None
+    d_bias: Optional[Tensor] = None
+    cpg = C // group
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:877]: should there be a check that it evenly divides?

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+    if output_mask[0]:
+        s = 1.0 / (HxW * cpg)
+        if gamma is not None:
+            ds_val = torch.mul(ds, gamma.reshape(1, C)).reshape(N, group, cpg).sum(2)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:881]: `gamma.unsqueeze(0)`, `unsqueeze` is typically easier for backends to reason about

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+            ds_val = ds.reshape(N, group, cpg).sum(2, dtype=acc_dtype)
+            db_val = db.reshape(N, group, cpg).sum(2, dtype=acc_dtype)
+            c1 = torch.mul(
+                rstd.reshape(N, group, 1),
+                torch.ones((1, group, cpg), dtype=acc_dtype, device=rstd.device),
+            )
+        c2 = (db_val * mean - ds_val) * rstd * rstd * rstd * s
+        c3 = -c2 * mean - db_val * rstd * s
+
+        c1 = c1.reshape(N, group, cpg, 1)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:897]: unsqueeze

File: torch/_decomp/decompositions.py [hunk 5]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+        d_input = (
+            torch.mul(grad_output.reshape(N, group, cpg, HxW), c1)
+            + torch.mul(input.reshape(N, group, cpg, HxW), c2)
+            + c3
+        )
+        d_input = d_input.reshape(input.shape).to(input.dtype)
+    if output_mask[1]:
+        d_gamma = ((ds.view(N, group, cpg) - db.view(N, group, cpg) * mean.view(N, group, 1)) * rstd.view(N, group, 1)).reshape(C)
+    if output_mask[2]:
+        d_bias = db.reshape(C)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:909]: but `db` has size (N,C)?

Comment (malfet) [reply torch/_decomp/decompositions.py:909]: You are right, it should be `db.sum(0)`

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py]: wait is `pw_cast_for_opmath` the right decorator to use here? Do you want to upcast all inputs to fp32?

Comment (ngimel) [reply torch/_decomp/decompositions.py]: idk how it's implemented in eager, but morally yes, all ops should be in fp32? E.g. in batch norm backward all intermediate computations are in fp32. 
We probably should be able to check that all inputs are of expected type, and type promotion decorations prevent us from doing so, but oh well.

File: torch/_decomp/decompositions.py [hunk 6]
```diff
@@ -847,6 +847,70 @@ def native_group_norm(
...
+    mean: Tensor,
+    rstd: Tensor,
+    gamma: Optional[Tensor],
+    N: int,
+    C: int,
+    HxW: int,
+    group: int,
+    output_mask: List[bool],
+) -> Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor]]:
+    assert grad_output.shape == input.shape
```
[the remark this answers is not in the document]
Comment (malfet) [reply torch/_decomp/decompositions.py]: Updated, not sure if this is what you've meant though...

File: torch/_decomp/decompositions.py [hunk 7]
```diff
@@ -847,6 +847,76 @@ def native_group_norm(
...
+    HxW: int,
+    group: int,
+    output_mask: List[bool],
+) -> Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor]]:
+    utils.check_same_device(
+        grad_output, input, mean, rstd, allow_cpu_scalar_tensors=False
+    )
+    utils.check_same_shape(input, grad_output, allow_cpu_scalar_tensors=False)
+    utils.check_same_shape(mean, rstd, allow_cpu_scalar_tensors=False)
+    assert input.numel() == N * C * HxW
```
Comment (Chillee) [inline torch/_decomp/decompositions.py:869]: I think @ngimel is suggesting to do something like https://github.com/pytorch/pytorch/blob/master/torch/_refs/__init__.py#L560

**Review by Chillee — APPROVED** [review]
LGTM, although I'd wait for Natalia's approval :P 

Would be nice to get the test in core somewhere.

Also, if you wanna try it, might be worth benchmarking how well Inductor lowers this op :)

https://github.com/pytorch/torchdynamo/pull/785

File: torch/_decomp/decompositions.py [hunk 8]
```diff
@@ -847,6 +847,76 @@ def native_group_norm(
...
+            torch.mul(grad_output.reshape(N, group, cpg, HxW), c1)
+            + torch.mul(input.reshape(N, group, cpg, HxW), c2)
+            + c3
+        )
+        d_input = d_input.reshape(input.shape).to(input.dtype)
+    if output_mask[1]:
+        d_gamma = (
+            (ds.view(N, group, cpg) - db.view(N, group, cpg) * mean.unsqueeze(-1))
+            * rstd.unsqueeze(-1)
+        ).reshape(C)
```
Comment (Chillee) [inline torch/_decomp/decompositions.py:913]: Actually, I saw some shape errors with this line when patching this decomp in and trying out @eellison 's per-operator benchmarks.

(from torchdynamo)

```
python ./benchmarks/microbenchmarks/operatorbench.py --op=aten.native_group_norm_backward.default --dtype=float16 --suite=timm
```

Comment (ngimel) [reply torch/_decomp/decompositions.py:913]: Yeah similar to `d_bias` there should be `.sum(0)` here

Comment (malfet) [reply torch/_decomp/decompositions.py]: Added `.sum(0)`, though it still need reshape here.

**Review by ngimel — APPROVED** [review]
Thanks!
